pandas/core/arrays/datetimes.py

```diff
@@ -1343,21 +1355,20 @@ def _generate_regular_range(cls, start, end, periods, freq):
                              "if a 'period' is given.")
 
         data = np.arange(b, e, stride, dtype=np.int64)
-        data = cls._simple_new(data.view(_NS_DTYPE), None, tz=tz)
     else:
         tz = None
         # start and end should have the same timezone by this point
-        if isinstance(start, Timestamp):
+        if start is not None:
             tz = start.tz
-        elif isinstance(end, Timestamp):
+        elif end is not None:
             tz = end.tz
 
         xdr = generate_range(start=start, end=end,
                              periods=periods, offset=freq)
 
-        values = np.array([x.value for x in xdr])
-        data = cls._simple_new(values, freq=freq, tz=tz)
+        data = np.array([x.value for x in xdr], dtype=np.int64)
 
+    data = cls._simple_new(data.view(_NS_DTYPE), freq=freq, tz=tz)
```

Comment:
its pretty arbitrary that you are viewing as M8[ns] rather than i8 here. let's be consistent (prob just i8 is fine), 
though I think this IS i8 already?

Reply:
Yah, there is a lot of casting back-and-forth.  I'll try to cut down on it.

General thought is that the values passed to `_simple_new` should already be in their correct forms (and master currently has some weird behavior, like passing a list in one case).  Sharing code between Datetime/Timedelta/Period pretty much requires that an exception be made for i8.